### DISCUSSION TITLE
LSP functionality in all files and not just repos and such

### DISCUSSION THREAD

**Discussion by JoshieKun** [LazyVim/LazyVim#2541] [General]
Hi, i'm tired of searching how to make LSP available for every file that supports its format i cant find anything about it. 

my problem is that is a enter a project with a github repo LSP work perfectly but if a try to create a new project with just one file of an LSP that i have installed this does not work at all it says that it does not have a attached LSP but that it support the file.

my config is lazy.vim with almost everything as factory default. 
using Mason and everything

for example im learning ruby rn if a try to create just a file to try some snippets of code. i cant get lsp and its really bothering me 
![image](https://github.com/LazyVim/LazyVim/assets/55514234/fd3602be-e7c0-4791-bea1-23c10cc4339a)

**Comment by dpetka2001** [comment]
Every LSP server behaves differently. You would have to go [here](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md) and search for the LSP server that you're interested in to see what the `root_dir` option expects. 

For example, for `Solargraph` you can see that `root_dir = root_pattern("Gemfile", ".git")`. That means that this LSP server expects either a `Gemfile` or a `.git` pattern to discover the `root_dir` and be able to attach. 

Some LSP servers support the option `single_file_support = true` (for example `clangd` or `pyright`), which means that you can simply create a file and the LSP server will attach to it. But that depends on the implementation of the LSP server itself.

**Reply by JoshieKun** [reply]
Thank you so much for your fast a very good response now i understand better how the setup work. 

but then if imposible to make it work with a single file if its not specified in the server itself ? 

if i tried to do something like this: 
```
	servers = {
				solargraph = {
					root_dir = root_pattern(".rb"),
				},
}
```
it just throw this error 

![image](https://github.com/LazyVim/LazyVim/assets/55514234/09a610af-4d61-4f87-b784-71973513a90d)

**Reply by dpetka2001** [reply]
Try to add at the top of your file before the line with `return` the following `local root_pattern = require("lspconfig.util").root_pattern` and try. I haven't done this myself, so not 100% sure if it'll work. But it also depends on if the server itself supports `single_file_support` or not. If it doesn't support it, then even if you defined it like that, it might not work. That is a question better asked in the Solargraph repo maybe? As they should know better about how the internals work.

**Reply by JoshieKun** [reply]
yes, you are right this is more a problem of solargraph than a lazy.vim one. thank you for fast response and help util information

**Reply by JoshieKun** [reply]
Idk if this posts are deleted or not. but if anyone wants  to recreate the functionality that i mention in this posts i found a solution AT LEAST FOR THIS SPECIFIC CASE with solargraph. big disclaimer tho, i have no idea if this will affect your performance so if your pc is not a powerful one do some testing.

and thanks @dpetka2001  for you help and guidance. i understand how the LSP work in nvim much better now 

```
servers = {
	solargraph = {
		root_dir = function(...)
		       return require("lspconfig.util").root_pattern("*.rb", ".git", "Gemfile")(...)
		end,
	},
}
```
